modules/library/jdbc/src/main/java/org/geotools/jdbc/JDBCDataStore.java

```diff
@@ -1549,70 +1547,201 @@ protected void insert(SimpleFeature feature, SimpleFeatureType featureType, Conn
-                        PreparedStatement ps = insertSQLPS( featureType, feature, keyValues, cx );
-                        try {
-                            ((PreparedStatementSQLDialect)dialect).onInsert(ps, cx, featureType);
-                            ps.execute();
-                        } finally {
-                            closeSafe( ps );
+                    Class binding = att.getType().getBinding();
+
+                    Object value = feature.getAttribute(colName);
+                    /*if (value == null) {
+                        if (!att.isNillable()) {
+                            //TODO: throw an exception
                         }
+                    }*/
+
+                    if (Geometry.class.isAssignableFrom(binding)) {
+                        Geometry g = (Geometry) value;
+                        int srid = getGeometrySRID(g, att);
+                        int dimension = getGeometryDimension(g, att);
+                        dialect.setGeometryValue(g, dimension, srid, binding, ps, i);
                     } else {
-                        String sql = insertSQL(featureType, feature, keyValues, cx);
-                        
-                        //TODO: execute in batch to improve performance?
-                        ((BasicSQLDialect)dialect).onInsert(st, cx, featureType);
-                        
-                        LOGGER.log(Level.FINE, "Inserting new feature: {0}", sql);
-                        st.execute(sql);
+                        dialect.setValue(value, binding, ps, i, cx);
                     }
-                    
-                    if ( keyValues == null ) {
-                        //grab the key values post insert
-                        keyValues = getLastValues(key,cx);
+                    if ( LOGGER.isLoggable( Level.FINE ) ) {
+                        LOGGER.fine( (i) + " = " + value );
                     }
-                    
-                    //report the feature id as user data since we cant set the fid
-                    String fid = featureType.getTypeName() + "." + encodeFID(keyValues);
-                    feature.getUserData().put("fid", fid);
+                    i++;
                 }
 
-                //st.executeBatch();
-            } catch (SQLException e) {
-                String msg = "Error inserting features";
-                throw (IOException) new IOException(msg).initCause(e);
-            } finally {
-                closeSafe(st);
+                keysFetcher.setKeyValues(dialect, ps, cx, featureType, feature, i);
+
+                dialect.onInsert(ps, cx, featureType);
+                ps.addBatch();
+            }
+            int[] inserts = ps.executeBatch();
+            checkAllInserted(inserts, features.size());
+            keysFetcher.postInsert(featureType, features, ps);
+        } finally {
+            closeSafe(ps);
+        }
+    }
+
+    private static void checkAllInserted(int[] inserts, int size) throws IOException {
```

Comment:
Because it's more a function than a method and it's more testable like that. By the way, I'v forgot to add the test. Fixing that.